tests/cxx/ptr_ref_aliases.cc

```diff
@@ -0,0 +1,182 @@
+//===--- ptr_ref_aliases.cc - test input file for iwyu --------------------===//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+//===----------------------------------------------------------------------===//
+
+// IWYU_ARGS: -Xiwyu --check_also=tests/cxx/ptr_ref_aliases-d1.h \
+//            -Xiwyu --check_also=tests/cxx/ptr_ref_aliases-d2.h \
+//            -I .
+
+// Tests various cases of type alias usages when the underlying type is
+// a pointer or a reference. Particularly, tests that `UseKind` parameter value
+// is correct.
+
+#include "tests/cxx/ptr_ref_aliases-d1.h"
+#include "tests/cxx/ptr_ref_aliases-d2.h"
+#include "tests/cxx/ptr_ref_aliases-d3.h"
+#include <typeinfo>
+
+NonProvidingPtrAlias GetPtr();
+NonProvidingRefAlias GetRef();
+
+NonProvidingPtrAlias ptr_nonprovided = GetPtr();
+NonProvidingRefAlias ref_nonprovided = GetRef();
+
+ProvidingPtrAlias ptr_provided = GetPtr();
+ProvidingRefAlias ref_provided = GetRef();
+
+struct Base {
+  virtual IndirectBase* RetPtr1();
+  virtual IndirectBase& RetRef1();
+  virtual IndirectBase* RetPtr2();
+  virtual IndirectBase& RetRef2();
+};
+
+struct Derived : Base {
+  // IWYU: Indirect is...*ptr_ref_aliases-i2.h
+  NonProvidingPtrAlias RetPtr1() override;
+  // IWYU: Indirect is...*ptr_ref_aliases-i2.h
+  NonProvidingRefAlias RetRef1() override;
+  ProvidingPtrAlias RetPtr2() override;
+  ProvidingRefAlias RetRef2() override;
+};
+
+void Fn() {
+  (void)sizeof(NonProvidingPtrAlias);
+  (void)sizeof(ptr_nonprovided);
+  // IWYU: Indirect is...*ptr_ref_aliases-i2.h
+  (void)sizeof(NonProvidingRefAlias);
+  // IWYU: Indirect is...*ptr_ref_aliases-i2.h
+  (void)sizeof(ref_nonprovided);
+  (void)sizeof(ProvidingRefAlias);
+  // TODO: avoid redundant reporting.
```

Comment:
Here and below -- what do you mean by "redundant" here?

It seems more like a proper bug; "redundant" makes me think it's an inconvenience and that the resulting IWYU suggestions are still correct. Could you maybe clarify these comments a little?

Reply:
"Unwanted" would be better? Or just "false-positive"? But these comments are removed in the second commit.

Reply:
Aha, I missed that 🙂. In that case it doesn't matter so much, but "unwanted" definitely seems better!